src/conn/conn_prefetch.c

```diff
@@ -95,6 +102,20 @@ __wt_prefetch_thread_run(WT_SESSION_IMPL *session, WT_THREAD *thread)
         TAILQ_REMOVE(&conn->pfqh, pe, q);
         --conn->prefetch_queue_count;
 
+        /*
+         * If the cache is getting close to its eviction clean trigger, don't attempt to pre-fetch
+         * the current ref as we may hang if the cache becomes full and we need to wait until space
+         * in the cache clears up. Repeat this process until either eviction has evicted enough
+         * eligible pages (allowing pre-fetch to read into the cache), or we iterate through and
+         * remove all the refs from the pre-fetch queue and pre-fetch becomes a no-op.
+         */
+        if (__wt_eviction_clean_pressure(session)) {
+            F_CLR_ATOMIC_8(pe->ref, WT_REF_FLAG_PREFETCH);
+            __wt_spin_unlock(session, &conn->prefetch_lock);
+            __wt_free(session, pe);
+            continue;
+        }
+
```

Comment:
Maybe a dumb question, but: do we want this at all if we have the timeout?

Reply:
My bad, I meant to put the timeout after this block of logic. I'll fix this now.

Reply:
OK, fair enough then. Thanks